Local Navigation: Improve style and behavior across screen sizes

ryelle opened the pull request:
This PR implements the style & behavior changes described in #535. The biggest change is that the navigation will now switch to the collapsed (mobile) view if it does not have enough space to stay on one single line (fixes https://github.com/WordPress/wporg-documentation-2022/issues/80). 

I did that by dropping in a second menu automatically - a filter is added to any nav blocks with `wporg/local-navigation-bar` as a parent block. That filter (`add_extra_navigation`) runs over the render_block HTML, so it can use all the same block attributes to create a second `wp:navigation` which is always in the collapsed (mobile) state. It's also given a special class name, so that it can be hidden/made visible by CSS. This ensures only one menu is visible/accessible at a time. In the javascript, a function is added to resize to check the available width against the section title & navigation items, and toggles the appropriate class based on what fits better.

Using a second menu like this also avoids trying to work around the current navigation menu toggle code, which has changed dramatically in the past.

Since the "mobile" view now shows on non-mobile screen sizes, the design was also updated and the styles here switch to a dropdown style when this version is shown at >600px wide.

There are also style changes to update focus styles & minor spacing fixes.

**Screenshots**

Documentation, using the page from https://github.com/WordPress/wporg-documentation-2022/issues/80:

| Default | Open menu | Mobile menu |
|---|---|---|
| ![](https://github.com/WordPress/wporg-mu-plugins/assets/541093/e6f21242-9fbc-4874-bd20-a80a9b64a629) | ![](https://github.com/WordPress/wporg-mu-plugins/assets/541093/b70a6a89-6982-4d72-9823-77ce22d89c3b) | ![](https://github.com/WordPress/wporg-mu-plugins/assets/541093/c9d22d5f-855d-4e8e-b9cb-5d04b22a3cb2) |

The bottom border should only show when scrolling:

https://github.com/WordPress/wporg-mu-plugins/assets/541093/c9225443-b58c-4be2-b07b-eafd621ca259

Here's an example of the menu collapsing when it would have wrapped (plus the "L3 header" variant updates):
https://github.com/WordPress/wporg-mu-plugins/assets/541093/71c069fd-9a7c-4c28-9e27-50c385faa286

Some long page titles still overflow the space available even with the collapsed nav menu. I've just set them to overflow:hidden, and at 550px wide, the page title is hidden.

| Full title visible | Overflow kicks in | Small screen, title hidden |
|---|---|---|
| ![](https://github.com/WordPress/wporg-mu-plugins/assets/541093/4b2bb17f-7e64-4a11-9752-01df7470a68a) | ![](https://github.com/WordPress/wporg-mu-plugins/assets/541093/3abb9b79-a1be-46a8-b4dc-538550a0d039) | ![](https://github.com/WordPress/wporg-mu-plugins/assets/541093/fa7f1297-ad4c-4ead-9bb2-51e7e1721335) |

**To test**

Check out this PR and use it with any site, it doesn't _require_ any local changes.

I'll have a separate PR for Developer to implement the different page level header templates, which was used to capture some of these screenshots.

ryelle committed:
Prevent menu from wrapping to second line

ryelle committed:
Use the block-specific hook to save a conditional

ryelle committed:
Move the custom property section to the top of the file

ryelle committed:
Style the mobile modal for medium & small screens, update focus states and spacing to match designs

ryelle committed:
Ensure child navigation block uses the expected attributes

ryelle committed:
Add an extra navigation block which is always collapsed, hidden by default

ryelle committed:
Adjust spacing & breakpoints

ryelle committed:
Update alignment of spacer dot

ryelle committed:
Hide the page title on small screens to prevent overflow

ryelle committed:
Prevent very long titles from wrapping on medium-small screens

ryelle committed:
Show a mobile menu (chevron dropdown) when the local nav and/or page title would cause the text to wrap

ryelle committed:
Get the gap value dynamically

On some sites (like Documentation), this is 24px, while it's 20px on Developer.

ryelle committed:
Ensure the focus style appears inside the overflow boundary

ryelle committed:
Fix alignment issues

ryelle committed:
Local navs with bottom borders should only show the border on scroll

StevenDufresne commented on mu-plugins/blocks/local-navigation-bar/src/view.js:
```diff
@@ -36,6 +36,64 @@ function init() {
+				paddingInlineEnd.replace( 'px', '' ) * 1 -
+				gap.replace( 'px', '' ) * 1;
+
+			const titleElement = container.querySelector( 'p.wp-block-site-title, div.wp-block-group' );
```
Noting that `p.wp-block-site-title` is specific and we would be then checking `getBoundingClientRect` on a `null` object.

ryelle replied:
Ah, because we're using that as the heading for showcase mobile… I assumed this would always be a `p` because the page's real `h1` appears in the content (this is true for Developer, Documentation, …everything but Showcase). I've asked about whether Showcase is an exception or another variation https://github.com/WordPress/wporg-mu-plugins/issues/535#issuecomment-1942110550

ryelle replied:
Addressed in https://github.com/WordPress/wporg-mu-plugins/pull/573/commits/9c8edd9ec319bfae2bc8a014cec6e56befbe6928

adamwoodnz commented:
Looking great, still testing, but found a broken width on Developer, looks like padding causing the wrapping

<img src="https://github.com/WordPress/wporg-mu-plugins/assets/1017872/31d164f8-051d-4391-8db3-1100fba5a59d" width="400">

adamwoodnz commented:
Seems unnecessary to me to show the title on the homepage on tablet and mobile, especially as the local nav isn't sticky on those screen sizes, so it isn't visible once scrolled

| Tablet | Mobile | Scrolled |
|-|-|-|
| ![developer wordpress org_(iPad)](https://github.com/WordPress/wporg-mu-plugins/assets/1017872/064132b6-68c0-44cb-bc2f-6cd91374dae8) | ![developer wordpress org_(Samsung Galaxy S20 Ultra)](https://github.com/WordPress/wporg-mu-plugins/assets/1017872/4a2651da-941d-4215-9c82-a53800174870) | ![developer wordpress org_(Samsung Galaxy S20 Ultra) (1)](https://github.com/WordPress/wporg-mu-plugins/assets/1017872/1d8fc2a8-87dc-481f-bbb9-9dd2bdc0cebf) |

adamwoodnz commented:
From #535 

> For level 3 and deeper pages, page name is hidden as breadcrumb takes that role. But once you scroll down, all pages, except for the section’s landing pages, show the page name.

I think we're missing this behaviour, for me the page name is visible in the local nav before scrolling

https://github.com/WordPress/wporg-mu-plugins/assets/1017872/c897c639-1582-4854-b41c-35db80bf3de3

adamwoodnz commented on mu-plugins/blocks/local-navigation-bar/src/view.js:
```diff
@@ -36,6 +36,64 @@ function init() {
+
+			const availableWidth =
+				window.innerWidth -
+				paddingInlineStart.replace( 'px', '' ) * 1 -
```
I have always used [`parseInt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt) for this type of conversion, not sure what's better, but maybe it's easier to understand the intention. Not a suggestion, just an illustration:
```suggestion
				parseInt(paddingInlineStart, 10) -
```

ryelle replied:
Personally I prefer `Number` over `parseInt` (the 10 always throws me off 🤷🏻), but this is a good point, I'll update it.

ryelle replied:
Ah, `parseInt` parses up to a non-number value, so I wouldn't need to remove the `'px'`… Okay, that's probably a good reason to use that here :)

ryelle replied:
Done in https://github.com/WordPress/wporg-mu-plugins/pull/573/commits/8e5458bd30923943a44636fefc5d1daa145090ed

adamwoodnz reviewed (approved):
This is great 👍 

Fixes the Docs bug for me and nails most of the new behaviour. I've left comments for things I've found but none are blockers imo.

ryelle commented:
> I think we're missing this behaviour, for me the page name is visible in the local nav before scrolling

I thought the animation of both the W and the page title was a little strange, so I'm waiting for some feedback on the main issue (see [the comment here](https://github.com/WordPress/wporg-mu-plugins/issues/535#issuecomment-1939906083)).

> Seems unnecessary to me to show the title on the homepage on tablet and mobile, especially as the local nav isn't sticky on those screen sizes, so it isn't visible once scrolled

Could you bring that up on the main issue? https://github.com/WordPress/wporg-mu-plugins/issues/535

ryelle committed:
Add class to fade in content on scroll

ryelle committed:
Use more broad title selector, check before using

ryelle committed:
Force section title elements to not wrap

ryelle committed:
Switch to parseInt for integer conversion

ryelle commented:
> Looking great, still testing, but found a broken width on Developer, looks like padding causing the wrapping

Fixed with https://github.com/WordPress/wporg-mu-plugins/pull/573/commits/bba37134b2e28db842aa9a00976b5b170485a6fb

> I think we're missing this behaviour, for me the page name is visible in the local nav before scrolling

I've added a class to fade-in the page title, and updated developer to use it. It'll look like this: 

https://github.com/WordPress/wporg-mu-plugins/assets/541093/84e87956-4a7b-48c6-b84d-e87506b38605

I think that addresses all the feedback here, but let me know if I missed something.

outdoor2kode reviewed (approved):
Tested and looks good 👍

ryelle committed:
Fix scrolling when mobile menu is longer than screen height

ryelle committed:
Fix height & dropdown spacing on mid-sized screens
